packages/cli/src/commands/hydrogen/env/pull.test.ts

```diff
@@ -0,0 +1,256 @@
+import {getConfig} from '../../../lib/shopify-config.js';
+import {linkStorefront} from '../link.js';
+
+import {pullVariables} from './pull.js';
+
+vi.mock('@shopify/cli-kit/node/ui', async () => {
+  const original = await vi.importActual<
+    typeof import('@shopify/cli-kit/node/ui')
+  >('@shopify/cli-kit/node/ui');
+  return {
+    ...original,
+    renderConfirmationPrompt: vi.fn(),
+  };
+});
+vi.mock('../link.js');
+vi.mock('../../../lib/admin-session.js');
+vi.mock('../../../lib/shopify-config.js');
+vi.mock('../../../lib/graphql.js', async () => {
+  const original = await vi.importActual<
+    typeof import('../../../lib/graphql.js')
+  >('../../../lib/graphql.js');
+  return {
+    ...original,
+    adminRequest: vi.fn(),
+  };
+});
+vi.mock('../../../lib/shop.js', () => ({
+  getHydrogenShop: () => 'my-shop',
+}));
+
+describe('pullVariables', () => {
+  const ADMIN_SESSION: AdminSession = {
+    token: 'abc123',
+    storeFqdn: 'my-shop',
+  };
+
+  beforeEach(async () => {
+    vi.mocked(getAdminSession).mockResolvedValue(ADMIN_SESSION);
+    vi.mocked(getConfig).mockResolvedValue({
+      storefront: {
+        id: 'gid://shopify/HydrogenStorefront/2',
+        title: 'Existing Link',
+      },
+    });
+    vi.mocked(adminRequest<PullVariablesSchema>).mockResolvedValue({
+      hydrogenStorefront: {
+        id: 'gid://shopify/HydrogenStorefront/1',
+        environmentVariables: [
+          {
+            id: 'gid://shopify/HydrogenStorefrontEnvironmentVariable/1',
+            key: 'PUBLIC_API_TOKEN',
+            value: 'abc123',
+            isSecret: false,
+          },
+          {
+            id: 'gid://shopify/HydrogenStorefrontEnvironmentVariable/1',
+            key: 'PRIVATE_API_TOKEN',
+            value: '',
+            isSecret: true,
+          },
+        ],
+      },
+    });
+  });
+
+  afterEach(() => {
+    vi.resetAllMocks();
+    mockAndCaptureOutput().clear();
+  });
+
+  it('makes a GraphQL call to fetch environment variables', async () => {
+    await inTemporaryDirectory(async (tmpDir) => {
+      await pullVariables({path: tmpDir});
+
+      expect(adminRequest).toHaveBeenCalledWith(
+        PullVariablesQuery,
+        ADMIN_SESSION,
+        {
+          id: 'gid://shopify/HydrogenStorefront/2',
+        },
+      );
+    });
+  });
+
+  it('writes environment variables to a .env file', async () => {
+    await inTemporaryDirectory(async (tmpDir) => {
+      const filePath = joinPath(tmpDir, '.env');
+
+      expect(await fileExists(filePath)).toBeFalsy();
+
+      await pullVariables({path: tmpDir});
+
+      expect(await readFile(filePath)).toStrictEqual(
+        'PUBLIC_API_TOKEN="abc123"\n' +
+          '# PRIVATE_API_TOKEN is marked as secret and its value is hidden\n' +
+          'PRIVATE_API_TOKEN=""\n',
+      );
+    });
+  });
+
+  it('warns if there are any variables marked as secret', async () => {
+    vi.mocked(adminRequest<PullVariablesSchema>).mockResolvedValue({
```

Comment:
The way you mock in-line here makes this test a lot more readable/reviewable because the whole context, variables, etc... are within the bounds of the 'it' callback. A lot less scrolling up and down.